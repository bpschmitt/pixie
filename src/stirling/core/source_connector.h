/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#pragma once

#include <string>
#include <vector>

#include "src/common/base/base.h"
#include "src/common/system/system.h"
#include "src/shared/types/types.h"
#include "src/stirling/core/connector_context.h"
#include "src/stirling/core/data_table.h"
#include "src/stirling/core/frequency_manager.h"

/**
 * These are the steps to follow to add a new data source connector.
 * 1. If required, create a new SourceConnector class.
 * 2. Add a new Create function with the following signature:
 *    static std::unique_ptr<SourceConnector> Create().
 *    In this function create an InfoClassSchema (vector of DataElement)
 * 3. Register the data source in the appropriate registry.
 */

namespace px {
namespace stirling {

class SourceConnector : public NotCopyable {
 public:
  SourceConnector() = delete;
  virtual ~SourceConnector() = default;

  /**
   * Initializes the source connector. Can only be called once.
   * @return Status of whether initialization was successful.
   */
  Status Init();

  /**
   * Sets the initial context for the source connector.
   * Used for context specific init steps (e.g. deploying uprobes on PIDs).
   */
  void InitContext(ConnectorContext* ctx);

  /**
   * Transfers all collected data to data tables.
   * @param ctx Shared context, e.g. ASID & tracked PIDs.
   * @param data_tables Map from the table number to DataTable objects.
   */
  void TransferData(ConnectorContext* ctx, const std::vector<DataTable*>& data_tables);

  /**
   * Pushes data in data tables into table store.
   */
  void PushData(DataPushCallback agent_callback, const std::vector<DataTable*>& data_tables);

  /**
   * Stops the source connector and releases any acquired resources.
   * May only be called after a successful Init().
   *
   * @return Status of whether stop was successful.
   */
  Status Stop();

  const std::string& name() const { return source_name_; }

  const ArrayView<DataTableSchema>& table_schemas() const { return table_schemas_; }

  static constexpr uint32_t TableNum(ArrayView<DataTableSchema> tables,
                                     const DataTableSchema& key) {
    uint32_t i = 0;
    for (i = 0; i < tables.size(); i++) {
      if (tables[i].name() == key.name()) {
        break;
      }
    }

    // Check that we found the index. This prevents compilation if name is not found,
    // during constexpr evaluation (which is awesome!).
    COMPILE_TIME_ASSERT(i != tables.size(), "Could not find name");

    return i;
  }

  /**
   * Utility function to convert time as recorded by in monotonic clock to real time.
   * This is especially useful for converting times from BPF, which are all in monotonic clock.
   */
  uint64_t ClockRealTimeOffset() const { return sysconfig_.ClockRealTimeOffset(); }

  virtual void SetDebugLevel(int level) { debug_level_ = level; }
  virtual void EnablePIDTrace(int pid) { pids_to_trace_.insert(pid); }
  virtual void DisablePIDTrace(int pid) { pids_to_trace_.erase(pid); }

  const FrequencyManager& sampling_freq_mgr() const { return sampling_freq_mgr_; }
  const FrequencyManager& push_freq_mgr() const { return push_freq_mgr_; }

 protected:
  explicit SourceConnector(std::string_view source_name,
                           const ArrayView<DataTableSchema>& table_schemas)
      : source_name_(source_name), table_schemas_(table_schemas) {}

  virtual Status InitImpl() = 0;

  // Provide a default InitContextImpl which does nothing.
  // SourceConnectors only need override if action is required on the initial context.
  virtual void InitContextImpl(ConnectorContext* /* ctx */) {}

  virtual void TransferDataImpl(ConnectorContext*, const std::vector<DataTable*>&) = 0;

  virtual Status StopImpl() = 0;

 protected:
  /**
   * Track state of connector. A connector's lifetime typically progresses sequentially
   * from kUninitialized -> kActive -> KStopped.
   *
   * kErrors is a special state to track a bad state.
   */
  enum class State { kUninitialized, kActive, kStopped, kErrors };

  // Sub-classes are allowed to inspect state.
  State state() const { return state_; }

  const system::Config& sysconfig_ = system::Config::GetInstance();

  FrequencyManager sampling_freq_mgr_;
  FrequencyManager push_freq_mgr_;

  // Debug members.
  int debug_level_ = 0;
  absl::flat_hash_set<int> pids_to_trace_;

 private:
  std::atomic<State> state_ = State::kUninitialized;

  const std::string source_name_;
  const ArrayView<DataTableSchema> table_schemas_;
};

}  // namespace stirling
}  // namespace px
