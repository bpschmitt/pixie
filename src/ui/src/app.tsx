/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  DARK_THEME, LIGHT_THEME, SnackbarProvider, VersionInfo,
} from 'app/components';
import Live from 'app/containers/App/live';
import PixieCookieBanner from 'configurable/cookie-banner';
import { LD_CLIENT_ID } from 'app/containers/constants';
import {
  Redirect, Route, Router, Switch,
} from 'react-router-dom';
import { makeCancellable, silentlyCatchCancellation } from 'app/utils/cancellable-promise';
import { isProd, PIXIE_CLOUD_VERSION } from 'app/utils/env';
import history from 'app/utils/pl-history';
import * as QueryString from 'query-string';

import {
  ThemeProvider, withStyles,
} from '@material-ui/core/styles';
import StyledEngineProvider from '@material-ui/core/StyledEngineProvider';
import { createStyles } from '@material-ui/styles';

import Axios from 'axios';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { CssBaseline } from '@material-ui/core';
import { withLDProvider } from 'launchdarkly-react-client-sdk';
import { AuthRouter } from 'app/pages/auth/auth';
import 'typeface-roboto';
import 'typeface-roboto-mono';
import { PixieAPIContext, PixieAPIContextProvider } from 'app/api';
import { AuthContextProvider, AuthContext } from 'app/common/auth-context';
import { EmbedContext, EmbedContextProvider } from 'app/common/embed-context';

// This side-effect-only import has to be a `require`, or else it gets erroneously optimized away during compilation.
require('./wdyr');

const RedirectWithArgs = (props) => {
  const {
    from,
    to,
    exact,
    location,
  } = props;

  return (
    <Redirect
      from={from}
      exact={exact}
      to={{ pathname: to, search: location.search }}
    />
  );
};

function useIsAuthenticated() {
  // Using an object instead of separate variables because using multiple setState does NOT batch if it happens outside
  // of React's scope (like resolved promises or Observable subscriptions). To make it atomic, have to use ONE setState.
  const [{ loading, authenticated, error }, setState] = React.useState({
    loading: true, authenticated: false, error: undefined,
  });

  const client = React.useContext(PixieAPIContext);
  React.useEffect(() => {
    if (!client) throw new Error('useIsAuthenticated needs to be called within a PixieAPIContextProvider!');

    setState({ loading: true, authenticated, error: undefined });
    const authPromise = makeCancellable(client.isAuthenticated());
    authPromise
      .then((isAuthenticated) => {
        setState({ loading: false, authenticated: isAuthenticated, error: undefined });
      })
      .catch(silentlyCatchCancellation)
      .catch((e) => {
        setState({ loading: false, authenticated: false, error: e });
      });

    return () => authPromise.cancel();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [client]);

  return { authenticated, loading, error };
}

export const App: React.FC = () => {
  const { authenticated, loading } = useIsAuthenticated();
  const { authToken } = React.useContext(AuthContext);

  const isEmbedded = window.location.pathname.startsWith('/embed');

  // If in an embedded environment, we need to wait until the authToken has been sent over from the parent.
  // While there is no authToken, we should not render the page, as all GQL requests will fail.
  if (isEmbedded && authToken.length === 0) {
    return null;
  }

  const authRedirectUri = window.location.pathname.length > 1
    ? encodeURIComponent(window.location.pathname + window.location.search)
    : '';
  const authRedirectTo = authRedirectUri ? `/login?redirect_uri=${authRedirectUri}` : '/login';

  return loading ? null : (
    <>
      <SnackbarProvider>
        <Router history={history}>
          <div className='center-content'>
            <Switch>
              <Route path='/auth' component={AuthRouter} />
              <RedirectWithArgs exact from='/login' to='/auth/login' />
              <RedirectWithArgs exact from='/logout' to='/auth/logout' />
              <RedirectWithArgs exact from='/signup' to='/auth/signup' />
              <RedirectWithArgs exact from='/auth-complete' to='/auth/cli-auth-complete' />
              {
                // 404s are handled within the Live route, after the user authenticates.
                // Logged out users get redirected to /login before the possibility of a 404 is checked.
                authenticated || isEmbedded ? <Route component={Live} />
                  : <Redirect from='/*' to={authRedirectTo} />
              }
            </Switch>
          </div>
        </Router>
        {!isProd() ? <VersionInfo cloudVersion={PIXIE_CLOUD_VERSION} /> : null}
      </SnackbarProvider>
      <PixieCookieBanner />
    </>
  );
};

// TODO(zasgar): Cleanup these styles. We probably don't need them all after we
// fix all of material styling.
const styles = () => createStyles({
  '@global': {
    '#root': {
      height: '100%',
      width: '100%',
    },
    html: {
      height: '100%',
    },
    body: {
      height: '100%',
      overflow: 'hidden',
      margin: 0,
      boxSizing: 'border-box',
    },
    ':focus': {
      outline: 'none !important',
    },
    // TODO(zasgar): remove this center content global.
    '.center-content': {
      height: '100%',
      width: '100%',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    },
  },
});

let StyledApp = withStyles(styles)(App);

if (LD_CLIENT_ID !== '') {
  StyledApp = withLDProvider({
    clientSideID: LD_CLIENT_ID,
  })(StyledApp);
}

const ThemedApp: React.FC = () => {
  const { setAuthToken } = React.useContext(AuthContext);
  const { setTimeArg } = React.useContext(EmbedContext);
  const [theme, setTheme] = React.useState<string | string[]>('dark');
  const [embedToken, setEmbedToken] = React.useState<string>('');

  // Parse query params to determine initial state of the page. These
  // params can also be set by the parent view, in an embedded context.
  React.useEffect(() => {
    const {
      theme: themeParam,
    } = QueryString.parse(window.location.search);

    if (themeParam) {
      setTheme(Array.isArray(themeParam) ? themeParam[0] : themeParam);
    }
  }, [setTheme]);

  // This is for an embedded environment.
  const listener = React.useCallback(async (event) => {
    const {
      data:
        {
          // embedPixieAPIKey is the auth token, and embedPixieAPIToken is a pixie API key.
          parentReady,
          embedPixieAPIKey,
          pixieTheme,
          pixieStartTime,
          embedPixieAPIToken,
        },
    } = event;

    if (pixieStartTime) {
      setTimeArg(pixieStartTime);
    }

    if (pixieTheme && (pixieTheme === 'light' || pixieTheme === 'dark')) {
      setTheme(pixieTheme);
    }

    // If the parent sends a ready message, it probably missed Pixie's
    // initial ready message. Send out another ready message, which
    // we know it will receive.
    if (parentReady) {
      window.top.postMessage({ pixieEmbedReady: true }, '*');
    }

    // In the embedded environment, we
    // expect authentication to be done using an auth token. The auth token
    // will be POSTed over, and is used to get a Pixie access token that is
    // attached to our requests using bearer auth.
    if (embedPixieAPIKey || embedPixieAPIToken) {
      // Only request a new access token if sent a new token.
      if (embedPixieAPIKey) {
        if (embedToken === embedPixieAPIKey) {
          return;
        }
        setEmbedToken(embedToken);
      }

      let response = null;
      try {
        response = await Axios.post('/api/auth/loginEmbedNew', {
          accessToken: embedPixieAPIKey ?? '',
          orgName: '',
        }, embedPixieAPIToken ? { headers: { 'pixie-api-key': embedPixieAPIToken } } : null);
      } catch (err) {
        return;
      }

      setAuthToken(response.data.token);
    }
  }, [embedToken, setEmbedToken, setAuthToken, setTimeArg]);

  React.useEffect(() => {
    window.addEventListener('message', listener);
    // Send a message to the parent frame to inform it that Pixie is listening
    // for postMessages. We use top, to send it to the topmost window.
    // window.postMessage is not enough for a child to contact the parent.
    window.top.postMessage({ pixieEmbedReady: true }, '*');
    return () => {
      window.removeEventListener('beforeunload', listener);
    };
  }, [listener]);

  let themeStyles = DARK_THEME;
  switch (theme) {
    case 'light':
      themeStyles = LIGHT_THEME;
      break;
    default:
  }
  return (
    <ThemeProvider theme={themeStyles}>
      <CssBaseline />
        <PixieAPIContextProvider apiKey=''>
          <StyledApp />
        </PixieAPIContextProvider>
    </ThemeProvider>
  );
};

ReactDOM.render(
  <StyledEngineProvider injectFirst>
    <AuthContextProvider>
      <EmbedContextProvider>
        < ThemedApp />
      </EmbedContextProvider>
    </AuthContextProvider>
  </StyledEngineProvider>, document.getElementById('root'));
