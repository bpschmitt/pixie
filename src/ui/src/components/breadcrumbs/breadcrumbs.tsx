/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as React from 'react';
import {
  Theme,
  makeStyles,
  ThemeProvider,
} from '@material-ui/core/styles';
import { createStyles } from '@material-ui/styles';
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';
import { Card, Popover } from '@material-ui/core';
import createTypography from '@material-ui/core/styles/createTypography';
import createSpacing from '@material-ui/core/styles/createSpacing';

import { CompletionItem } from 'app/components/autocomplete/completions';
import { Autocomplete } from 'app/components/autocomplete/autocomplete';
import useIsMounted from 'app/utils/use-is-mounted';
import { AutocompleteContext } from 'app/components/autocomplete/autocomplete-context';
import Paper from '@material-ui/core/Paper';

const useStyles = makeStyles(({
  spacing, typography, palette, breakpoints,
}: Theme) => createStyles({
  breadcrumbs: {
    height: '100%',
    scrollbarWidth: 'none', // Firefox
    '&::-webkit-scrollbar': {
      display: 'none',
    },
    [breakpoints.down('sm')]: {
      overflowX: 'scroll',
      display: 'flex',
    },
    display: 'flex',
    alignItems: 'center',
  },
  breadcrumb: {
    display: 'inline-flex',
    paddingRight: spacing(0.5),
    alignItems: 'center',
    height: '100%',
  },
  title: {
    fontWeight: typography.fontWeightMedium,
    paddingRight: spacing(0.5),
    paddingLeft: spacing(0.5),
    fontFamily: '"Roboto Mono", Monospace',
  },
  card: {
    width: '608px',
  },
  autocomplete: {
    maxHeight: '60vh',
  },
  value: {
    color: palette.primary.main,
    whiteSpace: 'nowrap',
    fontFamily: '"Roboto Mono", Monospace',
    maxWidth: spacing(50),
    overflowX: 'hidden',
    textOverflow: 'ellipsis',
  },
  selectable: {
    cursor: 'pointer',
    display: 'flex',
    alignItems: 'center',
  },
  body: {
    ...typography.body2,
    display: 'inline-block',
    color: palette.text.primary,
    height: spacing(5),
  },
  content: {
    display: 'flex',
    alignItems: 'center',
    paddingLeft: spacing(1),
    height: spacing(5),
  },
  dropdownArrow: {
    height: spacing(3),
  },
  spacer: {
    width: spacing(2),
  },
  input: {
    padding: `0 ${spacing(1)}px ${spacing(1)}px ${spacing(1)}px`,
  },
  completionsContainer: {
    // 80% as wide as the Command Input, and using 80% of the fontSize and 80% of the spacing
    maxWidth: '608px',
    maxHeight: '60vh',
  },
  // CSS for the front/back of the breadcrumb element.
  triangle: {
    position: 'absolute',
    height: spacing(3),
    width: spacing(3),
    overflow: 'hidden',
  },
  angle: {
    float: 'right', // Take the triangle out of the DOM, so it can overlap with other breadcrumbs.
  },
  tail: {
    width: spacing(2.4),
    height: spacing(3),
  },
  topArrow: {
    position: 'absolute',
    width: spacing(3),
    height: spacing(3),
    top: spacing(-1.5),
    right: spacing(-1.6),
    background: palette.foreground.grey3,
    transform: 'rotate(45deg)',
    borderRadius: spacing(0.4),
  },
  bottomArrow: {
    position: 'absolute',
    width: spacing(3),
    height: spacing(3),
    bottom: spacing(-1.5),
    right: spacing(-1.6),
    background: palette.foreground.grey3,
    transform: 'rotate(45deg)',
    borderRadius: spacing(0.4),
  },
  arrow: {
    position: 'absolute',
    width: spacing(3),
    height: spacing(3),
    left: spacing(-1.5),
    background: palette.foreground.grey3,
    transform: 'rotate(45deg)',
    borderRadius: spacing(0.4),
  },
  inputItem: {
    '&:hover': {
      backgroundColor: 'initial',
    },
    '&.Mui-focusVisible': {
      backgroundColor: 'initial',
    },
  },
  separator: {
    display: 'flex',
    alignItems: 'center',
    color: palette.foreground.one,
    fontWeight: 1000,
    width: spacing(1),
  },
  icon: {
    minWidth: spacing(4),
  },
  divider: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: palette.foreground.three,
    height: '75%',
    width: '1px',
    opacity: 0.4,
  },
}), { name: 'Breadcrumbs' });

export interface DialogDropdownProps {
  onSelect: (input: string) => void;
  getListItems: (input: string) => Promise<BreadcrumbListItem[]>;
  onClose: () => void;
  anchorEl: HTMLElement;
  placeholder: string;
  explanation?: React.ReactNode;
}

const useDialogStyles = makeStyles((theme) => createStyles({
  card: {
    width: '608px',
  },
  autocomplete: {
    maxHeight: '60vh',
  },
  completionsContainer: {
    // 80% as wide as the Command Input, and using 80% of the fontSize and 80% of the spacing
    maxWidth: '608px',
    maxHeight: '60vh',
  },
  explanationContainer: {
    backgroundColor: theme.palette.foreground.grey3,
    padding: theme.spacing(2),
    borderTop: 'solid 1px',
    borderColor: theme.palette.background.three,
  },
}));

export const DialogDropdown: React.FC<DialogDropdownProps> = ({
  placeholder,
  onSelect,
  onClose,
  getListItems,
  anchorEl,
  explanation,
}) => {
  const classes = useDialogStyles();

  const mounted = useIsMounted();
  const { allowTyping, requireCompletion } = React.useContext(AutocompleteContext);
  const inputRef = React.useRef<HTMLInputElement>(null);

  if (anchorEl && inputRef.current) {
    inputRef.current?.focus();
  }

  const onCompletionSelected = React.useCallback(
    (itemValue: string) => {
      if (requireCompletion) {
        // Completion required, refresh the valid options and verify that the user picked one of them.
        // As getListItems is async, we refresh it so that the check is run atomically and can't be out of date.
        if (typeof getListItems !== 'function') {
          throw new Error(
            `List items not gettable when selecting "${itemValue}"!`,
          );
        }
        getListItems(itemValue).then((items) => {
          if (items.length < 1) throw new Error('Failed to match the input to a valid choice!');
          if (mounted.current) {
            onSelect(itemValue);
          }
          // These two conditionals ARE separate: onSelect is able to cause this component to unmount before onClose.
          // If we called onClose anyway, React could try to update state on an unmounted component, which is an error.
          if (mounted.current) {
            onClose();
          }
        });
      } else {
        // Completion not required, so we don't care if the user picked an item that actually exists. Skip validation.
        onSelect(itemValue);
        onClose();
      }
    },
    // `mounted` is not in this array because changing it should ONLY affect whether the resolved promise acts.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [requireCompletion, getListItems, onClose, onSelect],
  );

  const getCompletions = React.useCallback(
    (input: string) => {
      if (typeof getListItems === 'function') {
        return getListItems(input).then((items) => {
          // TODO(nick,PC-630): This should be done on the API side. Once it is, remove this defensive coding.
          const seenNames = new Set<string>();
          const deduped = items.filter((item) => {
            if (seenNames.has(item.value)) {
              return false;
            }
            seenNames.add(item.value);
            return true;
          });
          const mapped: CompletionItem[] = deduped.map((item) => ({
            type: 'item',
            id: item.value,
            description: item.description,
            icon: item.icon,
            title: item?.title ?? item.value,
            autoSelectPriority: item.autoSelectPriority ?? 0,
          }));

          // If we don't require that the final result be one of the suggested items,
          // then add the user's current input as the first option, unless they have
          // already typed an existing option.
          if (!requireCompletion && !seenNames.has(input)) {
            mapped.unshift({
              type: 'item',
              id: input,
              description: '',
              icon: '',
              title: input,
              autoSelectPriority: 1,
            });
          }

          return mapped;
        });
      }

      return Promise.resolve([]);
    },
    [getListItems, requireCompletion],
  );

  // Used to shrink the <Autocomplete/>'s fonts and negative space, to not be as huge / central as the Command Input.
  const compactThemeBuilder = React.useMemo<(theme: Theme) => Theme>(
    () => (theme: Theme) => ({
      ...theme,
      typography: createTypography(theme.palette, {
        fontSize: theme.typography.fontSize * 0.8,
      }),
      spacing: createSpacing((factor) => theme.spacing(factor * 0.8)),
    }),
  [],
  );

  return (
    <Popover
      classes={{ paper: classes.completionsContainer }}
      anchorEl={anchorEl}
      keepMounted
      open={Boolean(anchorEl)}
      onClose={onClose}
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'left',
      }}
      transformOrigin={{
        vertical: 'top',
        horizontal: 'left',
      }}
      elevation={0}
    >
      <ThemeProvider theme={compactThemeBuilder}>
        <Card className={classes.card}>
          <AutocompleteContext.Provider
            value={{
              allowTyping,
              requireCompletion,
              inputRef,
              onOpen: 'clear',
              hidden: !anchorEl,
            }}
          >
            <Autocomplete
              className={classes.autocomplete}
              placeholder={placeholder}
              onSelection={onCompletionSelected}
              getCompletions={getCompletions}
            />
          </AutocompleteContext.Provider>
          {
            explanation != null
            && (
              <div className={classes.explanationContainer}>
                { explanation }
              </div>
            )
          }
        </Card>
      </ThemeProvider>
    </Popover>
  );
};

export interface BreadcrumbProps {
  title: string;
  value: string;
  selectable: boolean;
  getListItems?: (input: string) => Promise<Array<BreadcrumbListItem>>;
  onSelect?: (input: string) => void;
  omitKey?: boolean;
  placeholder?: string;
  explanation?: React.ReactElement;
}

const Breadcrumb: React.FC<BreadcrumbProps> = ({
  title,
  value,
  selectable,
  getListItems,
  onSelect,
  omitKey,
  placeholder,
  explanation,
}) => {
  const classes = useStyles();
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);

  const handleClick = React.useCallback(
    (event: React.MouseEvent<HTMLElement>) => {
      setAnchorEl(event.currentTarget.parentElement);
    },
    [setAnchorEl],
  );

  const onClose = React.useCallback(() => {
    setAnchorEl(null);
    // The input from the autocomplete retains focus despite being hidden otherwise.
    (document.activeElement as HTMLElement)?.blur();
  }, [setAnchorEl]);

  return (
    <div className={classes.breadcrumb}>
      <div className={classes.body}>
        <div className={classes.content}>
          {!omitKey && <span className={classes.title}>{`${title}: `}</span>}
          <span className={selectable ? classes.selectable : ''} onClick={selectable ? handleClick : null}>
            <span className={classes.value} >{value}</span>
            {selectable && (
              <div className={classes.dropdownArrow}>
                <ArrowDropDownIcon />
              </div>
            )}
          </span>
          {!selectable && <div className={classes.spacer} />}
          <DialogDropdown
            placeholder={placeholder || 'Filter...'}
            onSelect={onSelect}
            onClose={onClose}
            getListItems={getListItems}
            anchorEl={anchorEl}
            explanation={explanation}
          />
        </div>
      </div>
    </div>
  );
};

export interface BreadcrumbListItem {
  value: string;
  title?: string;
  description?: string;
  icon?: React.ReactNode;
  /** @see{CompletionItem} */
  autoSelectPriority?: number;
}

export interface BreadcrumbOptions extends BreadcrumbProps {
  allowTyping?: boolean;
  requireCompletion?: boolean;
  divider?: boolean;
}

export interface BreadcrumbsProps {
  breadcrumbs: BreadcrumbOptions[];
}

export const Breadcrumbs: React.FC<BreadcrumbsProps> = ({ breadcrumbs }) => {
  const classes = useStyles();
  // In case a breadcrumb doesn't override, give it the nearest context's values.
  const { allowTyping, requireCompletion } = React.useContext(
    AutocompleteContext,
  );
  return (
    <Paper elevation={2}>
      <div className={classes.breadcrumbs}>
        {breadcrumbs.map((breadcrumb, i) => (
          // Key is needed to prevent a console error when a key is missing in a list element.
          <AutocompleteContext.Provider
            key={`i-${breadcrumb.title}`}
            value={{
              allowTyping: breadcrumb.allowTyping ?? allowTyping,
              requireCompletion:
                breadcrumb.requireCompletion ?? requireCompletion,
              onOpen: 'clear',
            }}
          >
            <Breadcrumb key={i} {...breadcrumb} />
            {
              breadcrumb.divider && i !== breadcrumbs.length - 1 && <div className={classes.divider} />
            }
          </AutocompleteContext.Provider>
        ))}
      </div>
    </Paper>
  );
};
