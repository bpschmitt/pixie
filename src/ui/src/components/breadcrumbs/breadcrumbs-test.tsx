/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { screen, render } from '@testing-library/react';
import { ThemeProvider } from '@material-ui/core/styles';
import { DARK_THEME } from 'app/components';
import { Breadcrumbs } from './breadcrumbs';

describe('<Breadcrumbs/>', () => {
  it('renders correctly', async () => {
    const breadcrumbs = [
      {
        title: 'cluster',
        value: 'gke-prod',
        selectable: true,
        allowTyping: false,
        getListItems: async (input) => {
          if (input) {
            return [
              { value: 'cluster1' },
              { value: 'cluster2' },
              { value: 'cluster3' },
            ];
          }
          return [
            { value: 'cluster1' },
            { value: 'cluster2' },
            { value: 'cluster3' },
          ];
        },
      },
      {
        title: 'pod',
        value: 'pod-123',
        selectable: true,
        allowTyping: true,
        getListItems: async (input) => {
          if (input === '') {
            return [{ value: 'pod1' }, { value: 'pod2' }];
          }
          return [
            { value: 'some pod' },
            { value: 'another pod' },
            { value: 'pod' },
          ];
        },
      },
      {
        title: 'script',
        value: 'px/pod',
        selectable: false,
      },
    ];

    const { container } = render(
      <ThemeProvider theme={DARK_THEME}>
        <Breadcrumbs breadcrumbs={breadcrumbs} />
      </ThemeProvider>,
    );
    await screen.findByText('cluster:'); // Entirely to wait for actions
    expect(container).toMatchSnapshot();
  });
});
