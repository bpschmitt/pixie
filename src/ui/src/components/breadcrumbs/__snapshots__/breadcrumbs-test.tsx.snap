// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs/> renders correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1459qpu-MuiPaper-root"
  >
    <div
      class="Breadcrumbs-breadcrumbs-1"
    >
      <div
        class="Breadcrumbs-breadcrumb-2"
      >
        <div
          class="Breadcrumbs-body-8"
        >
          <div
            class="Breadcrumbs-content-9"
          >
            <span
              class="Breadcrumbs-title-3"
            >
              cluster: 
            </span>
            <span
              class="Breadcrumbs-selectable-7"
            >
              <span
                class="Breadcrumbs-value-6"
              >
                gke-prod
              </span>
              <div
                class="Breadcrumbs-dropdownArrow-10"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m7 10 5 5 5-5z"
                  />
                </svg>
              </div>
            </span>
          </div>
        </div>
      </div>
      <div
        class="Breadcrumbs-breadcrumb-2"
      >
        <div
          class="Breadcrumbs-body-8"
        >
          <div
            class="Breadcrumbs-content-9"
          >
            <span
              class="Breadcrumbs-title-3"
            >
              pod: 
            </span>
            <span
              class="Breadcrumbs-selectable-7"
            >
              <span
                class="Breadcrumbs-value-6"
              >
                pod-123
              </span>
              <div
                class="Breadcrumbs-dropdownArrow-10"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m7 10 5 5 5-5z"
                  />
                </svg>
              </div>
            </span>
          </div>
        </div>
      </div>
      <div
        class="Breadcrumbs-breadcrumb-2"
      >
        <div
          class="Breadcrumbs-body-8"
        >
          <div
            class="Breadcrumbs-content-9"
          >
            <span
              class="Breadcrumbs-title-3"
            >
              script: 
            </span>
            <span
              class=""
            >
              <span
                class="Breadcrumbs-value-6"
              >
                px/pod
              </span>
            </span>
            <div
              class="Breadcrumbs-spacer-11"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
